src/compiler/program.ts

```diff
@@ -2532,8 +2532,8 @@ namespace ts {
             }
 
             if (options.declarationDir) {
-                if (!options.declaration) {
-                    createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "declarationDir", "declaration");
+                if (!(options.declaration || options.composite)) {

```

Comment:
there's a function `getEmitDeclarations` that does the same and is used in all other places